compiler/noirc_frontend/src/elaborator/patterns.rs

```diff
@@ -401,12 +401,45 @@ impl<'context> Elaborator<'context> {
     pub(super) fn elaborate_variable(
         &mut self,
         variable: Path,
-        generics: Option<Vec<Type>>,
+        unresolved_turbofish: Option<Vec<UnresolvedType>>,
     ) -> (ExprId, Type) {
         let span = variable.span;
         let expr = self.resolve_variable(variable);
         let definition_id = expr.id;
 
+        let definition = self.interner.try_definition(definition_id);
+
+        // Resolve any generics if we the variable we have resolved is a function
+        // and if the turbofish operator was used.
+        let generics = if let Some(definition) = definition {
+            match &definition.kind {
+                DefinitionKind::Function(function) => {
+                    // Resolve generics using the expected kinds of the function we are calling
+                    let direct_generics =
+                        self.interner.function_meta(function).direct_generics.clone();
+
+                    unresolved_turbofish.map(|option_inner| {
+                        if option_inner.len() != direct_generics.len() {
+                            let type_check_err = TypeCheckError::IncorrectTurbofishGenericCount {
+                                expected_count: direct_generics.len(),
+                                actual_count: option_inner.len(),
+                                span,
+                            };
+                            self.push_err(type_check_err);
+                        }
+
+                        let generics_with_types = direct_generics.iter().zip(option_inner);
+                        vecmap(generics_with_types, |(generic, unresolved_type)| {
+                            self.resolve_type_inner(unresolved_type, &generic.kind)
+                        })
+                    })
```

Comment:
This code chunk seems to be repeated for method calls as well - we should have a helper for it

Reply:
I have made a helper `resolve_turbofish_generics`